File: 525-contiguous-array/contiguous-array.java
class Solution {
    public int findMaxLength(int[] nums) {
        Map map = new HashMap<>();
        int maxLength = 0, sum = 0;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i] == 0 ? -1 : 1;
            if (sum == 0) {
                maxLength = Math.max(maxLength, i + 1);
            } else if (map.containsKey(sum)) {
                maxLength = Math.max(maxLength, i - (int) map.get(sum));
            } else {
                map.put(sum, i);
            }
        }
        return maxLength;
    }
}

File: 525-contiguous-array/contiguous-array.js
var findMaxLength = function(nums) {
    let maxLen = 0;
    let count = 0;
    let counts = new Map();
    counts.set(0, -1);
    for (let i = 0; i < nums.length; i++) {
        count += nums[i] === 1 ? 1 : -1;
        if (counts.has(count)) {
            maxLen = Math.max(maxLen, i - counts.get(count));
        } else {
            counts.set(count, i);
        }
    }
    return maxLen;
};